### ISSUE TITLE
新转化的PaddleHubModule使用serving部署后图片参数如何传递？

### ISSUE THREAD

**Issue by Payne-Zhu** [PaddlePaddle/PaddleFormers#695]
使用resnet_v2_50_imagenet迁移学习后新产生的模型按照下列方式转化为PaddleHubModule；
再使用hub install 安装到本机后新模型xxx_test使用hub serving start -m xxx_test部署可以成功；
调用这个服务参数要怎么写？使用示例中的方法会出错；返回需要使用json数据(arnning: "This usage is out of date, please use 'application/json' as content-type to post to )，但是使用json的话图片如何转换传递呢？
如何实现原版的 resnet_v2_50_imagenet 通过 http://localhost:8866/predict/image/resnet_v2_50_imagenet 
传递Post参数image: data:image/jpeg;base64,... 就能直接返回的方式呢？

`import os
import numpy as np
import paddlehub as hub
from paddlehub.module.module import moduleinfo, serving

@moduleinfo(
    name="xxx_test",
    version="1.0.0",
    summary="ERNIE tiny which was fine-tuned on the chnsenticorp dataset.",
    author="payne",
    author_email="payne@qq.com",
    type="cv/xxx_model",
)
class RestNetFinetuned(hub.Module):
    def _initialize(self,
                    ckpt_dir="xxx_model",
                    num_class=4,
                    max_seq_len=128,
                    use_gpu=False,
                    batch_size=4):
        self.ckpt_dir = os.path.join(self.directory, ckpt_dir)

        self.module = hub.Module(name="resnet_v2_50_imagenet")
        inputs, outputs, program = self.module.context(trainable=True)

        self.vocab_path = self.module.get_vocab_path()

        reader = hub.reader.ImageClassificationReader(
            image_width=self.module.get_expected_image_width(),
            image_height=self.module.get_expected_image_height(),
            images_mean=self.module.get_pretrained_images_mean(),
            images_std=self.module.get_pretrained_images_std(),
        )

        pooled_output = outputs["feature_map"]

        feed_list = [inputs["image"].name]

        config = hub.RunConfig(
            use_data_parallel=False,
            use_cuda=use_gpu,
            batch_size=batch_size,
            checkpoint_dir=self.ckpt_dir,
            strategy=hub.finetune.strategy.DefaultFinetuneStrategy())

        self.cls_task = hub.ImageClassifierTask(
            data_reader=reader,
            feature=pooled_output,
            feed_list=feed_list,
            num_classes=4,
            config=config)

    @serving
    def predict(self, data, return_result=False, accelerate_mode=True):
        run_states = self.cls_task.predict(data=data)
        # return run_states
        results = [run_state.run_results for run_state in run_states]
        prediction = []
        for batch_result in results:
            # get predict index
            batch_result = np.argmax(batch_result, axis=2)[0]
            batch_result = batch_result.tolist()
            prediction += batch_result
        return prediction
        

if __name__ == "__main__":
    dm_4_npc = RestNetFinetuned(ckpt_dir="dm_4_npc")
    data = ["data/1.png","data/4.png","data/2.png","data/3.png"]
    run_states = dm_4_npc.predict(data=data)
    print(run_states)`

**Comment by grasswolfs** [comment]
> 使用resnet_v2_50_imagenet迁移学习后新产生的模型按照下列方式转化为PaddleHubModule；
> 再使用hub install 安装到本机后新模型xxx_test使用hub serving start -m xxx_test部署可以成功；
> 调用这个服务参数要怎么写？使用示例中的方法会出错；返回需要使用json数据(arnning: "This 

请问下payne@qq.com是您的邮箱么？可以沟通下更细节的背景么，也好更好的为您提供技术支持。
方便的话是否可以加个微信沟通下？微信号：282214037

**Comment by Steffy-zxf** [comment]
@ittaoli 你好！将Finetune后模型参数转成PaddleHub module，并可以使用Hub Serving部署。可以参考 回复：https://github.com/PaddlePaddle/PaddleHub/issues/692。

图片在网络中如何传输：可以先将图片使用cv2读入，之后将读入的图片数据(`ndarray`)转化为`base64`格式。即可在网路中传输。

如何使用Hub Serving 部署服务：
 首先，启动服务段：
```shell
$ hub serving start -m pyramidbox_lite_server_mask
```

之后，客户端请求：
```python
import requests
import json
import cv2
import base64

def cv2_to_base64(image):
    data = cv2.imencode('.jpg', image)[1]
    return base64.b64encode(data.tostring()).decode('utf8')

# 发送HTTP请求
data = {'data':[cv2_to_base64(cv2.imread("/PATH/TO/IMAGE"))]}
headers = {"Content-type": "application/json"}
url = "http://127.0.0.1:8866/predict/xxx_test"
r = requests.post(url=url, headers=headers, data=json.dumps(data))

# 打印预测结果
print(r.json()["results"])
```

**Comment by Payne-Zhu** [comment]
@Steffy-zxf 你好；按照您提供的方法返回结果为 
{"msg":"Please check data format!","results":"","status":"-1"}
依然不能正常调用，服务端显示错误信息：
[Errno 2] No such file or directory: '/9j/4AAQSkZJRgABAQ...
感觉是服务端读到了发送的base64图片数据；但是把他当成图片路径去用了，是不是我少写了什么转换函数？

**Comment by Steffy-zxf** [comment]
感谢反馈！由于图像分类`ImageClassifierTask.predict`接口接受的是图片路径，所以图像分类任务使用serving确实有以上问题。这个问题之后将会修复。修复后，第一时间告知您。

**Comment by FourierX9** [comment]
你好! 我们在做口罩人脸检测 (地址:http://127.0.0.1:8866/predict/image/pyramidbox_lite_server_mask) 的时候,也遇到了相似的情况.paddlehub版本是1.7.1.图片发送用的代码段是上面转成Base64的方法.收到sever返回的结果是{"msg":"Please check data format!","results":"","status":"-1"}. 服务端报了:serving_method() missing 1 required positional argument: 'images'

附上我的代码:
```Python
# coding: utf8
import requests
import json
import base64
import os
import cv2

def cv2_to_base64(image):
    data = cv2.imencode('.jpg', image)[1]
    return base64.b64encode(data.tostring()).decode('utf8')

# 发送HTTP请求
data = {'data': [cv2_to_base64(cv2.imread("test_mask_detection.jpg"))]}
headers = {"Content-type": "application/json"}
url = "http://192.200.125.116:8866/predict/image/pyramidbox_lite_server_mask"
# 指定检测方法为pyramidbox_lite_server_mask并发送post请求
r = requests.post(url=url, headers=headers, data=json.dumps(data))

results = eval(r.json()["results"])

# 保存检测生成的图片到output文件夹，打印模型输出结果
if not os.path.exists("output"):
    os.mkdir("output")
for item in results:
    with open(os.path.join("output", item["path"]), "wb") as fp:
        fp.write(base64.b64decode(item["base64"].split(',')[-1]))
        item.pop("base64")
print(json.dumps(results, indent=4, ensure_ascii=False))

**Comment by codeWang-J** [comment]
收到


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "PaddlePaddle/PaddleHub"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2020年11月24日(星期二) 下午4:39
收件人:&nbsp;"PaddlePaddle/PaddleHub"<PaddleHub@noreply.github.com&gt;;
抄送:&nbsp;"Subscribed"<subscribed@noreply.github.com&gt;;
主题:&nbsp;Re: [PaddlePaddle/PaddleHub] 新转化的PaddleHubModule使用serving部署后图片参数如何传递？ (#695)





 
你好! 我们在做口罩人脸检测 (地址:http://127.0.0.1:8866/predict/image/pyramidbox_lite_server_mask) 的时候,也遇到了相似的情况.paddlehub版本是1.7.1.图片发送用的代码段是上面转成Base64的方法.收到sever返回的结果是{"msg":"Please check data format!","results":"","status":"-1"}. 服务端报了:serving_method() missing 1 required positional argument: 'images'
 
附上我的代码:
 # coding: utf8 import requests import json import base64 import os import cv2 def cv2_to_base64(image):     data = cv2.imencode('.jpg', image)[1]     return base64.b64encode(data.tostring()).decode('utf8') # 发送HTTP请求 data = {'data': [cv2_to_base64(cv2.imread("test_mask_detection.jpg"))]} headers = {"Content-type": "application/json"} url = "http://192.200.125.116:8866/predict/image/pyramidbox_lite_server_mask" # 指定检测方法为pyramidbox_lite_server_mask并发送post请求 r = requests.post(url=url, headers=headers, data=json.dumps(data)) results = eval(r.json()["results"]) # 保存检测生成的图片到output文件夹，打印模型输出结果 if not os.path.exists("output"):     os.mkdir("output") for item in results:     with open(os.path.join("output", item["path"]), "wb") as fp:         fp.write(base64.b64decode(item["base64"].split(',')[-1]))         item.pop("base64") print(json.dumps(results, indent=4, ensure_ascii=False))

**Comment by FourierX9** [comment]
我现在的方法是把PaddleHub的版本降到v1.5.4,先用起来了

**Comment by haoyuying** [comment]
Since you haven't replied for more than 3 months, we have closed this issue/pr.
If the problem is not solved or there is a follow-up one, please reopen it at any time and we will continue to follow up.
由于您超过三个月未回复，我们将关闭这个issue/pr。
若问题未解决或有后续问题，请随时重新打开，我们会继续跟进。
